### PR TITLE
refactor(common): move low level utilities to @endo/common

### PR DESCRIPTION
**Pull request by erights**
closes: #XXXX
refs: #1815

## Description

Successor to https://github.com/endojs/endo/pull/1815 with utilities moved to a new @endo/utils package.

Note: I did not label this `refactor(utils)!: ...` because I intend to add reexports to keep compat before declaring this ready for review.

### CODE REVIEW

File: packages/patterns/index.js [hunk 1]
```diff
@@ -71,9 +71,5 @@ export {
   kindOf,
 } from './src/patterns/patternMatchers.js';
 
-// ////////////////// Temporary, until these find their proper home ////////////
-
-export { listDifference, objectMap } from './src/utils.js';
```
Comment (erights) [inline packages/patterns/index.js:76]: Reexport these as deprecated, with a suggestion to import directly from @endo/utils. That would preserve compat for this PR.

**Comment by kriskowal** [comment]
I want this to be called `@endo/score` as a pun on `underscore`.

To allow manual minimization of the weight of a program’s dependency graph, each utility function should be separately importable, like `@endo/score/list-difference.js`.

File: packages/utils/CHANGELOG.md
Comment (kriskowal) [inline packages/utils/CHANGELOG.md:1]: This file can be deleted. We should revise `create-package.sh` to do so.

File: packages/utils/README.md
Comment (kriskowal) [inline packages/utils/README.md:1]: A minimal README would be good.

File: packages/utils/package.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "@endo/utils",
+  "version": "1.0.1",
+  "private": true,
+  "description": null,
```
Comment (kriskowal) [inline packages/utils/package.json:5]: A description would be good.

File: packages/utils/test/test-index.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { test } from './prepare-test-env-ava.js';
+
+test('place holder', async t => {
+  t.pass();
+});
```
Comment (kriskowal) [inline packages/utils/test/test-index.js:5]: Should be replaced with tests.

**Comment by erights** [comment]
> I want this to be called `@endo/score` as a pun on `underscore`.

No single entendre puns. I'll allow it if you also have some rationalization, however bizarre or awkward, why "score" is a sensible name for this package. ;)

**Comment by kriskowal** [comment]
> > I want this to be called `@endo/score` as a pun on `underscore`.
> 
> No single entendre puns. I'll allow it if you also have some rationalization, however bizarre or awkward, why "score" is a sensible name for this package. ;)

I can only appeal to the tradition of:

- https://jquery.com/ for establishing the use of a single symbol (`$`) for a utility module
- https://underscorejs.org/ for establishing the use of (`_`) in particular as a utility module name
- https://lodash.com/ for riffing on synonyms of underscore (`_`)
- endo and under sounding similar

However, I cannot think of a winning pun on “score” that cuts it, which is a strike that really underlines your point. Scratch that.

**Comment by erights** [comment]
> However, I cannot think of a winning pun on “score” that cuts it, which is a strike that really underlines your point. Scratch that.

I asked for a double entendre and got a meta entendre. Serves me right ;)

File: packages/patterns/index.js (hunk 1 above)
Comment (erights) [reply packages/patterns/index.js:76]: Done

File: packages/utils/CHANGELOG.md
Comment (erights) [reply packages/utils/CHANGELOG.md:1]: I deleted this occurrence, but I leave changes to `create-package.sh` up to you.

Resolving.

**Comment by erights** [comment]
On a slack channel, we settled on a package name of `common`

File: packages/utils/package.json (hunk 2 above)
Comment (erights) [reply packages/utils/package.json]: Done

File: packages/utils/test/test-index.js (hunk 3 above)
Comment (erights) [reply packages/utils/test/test-index.js]: Done!

File: packages/common/index.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export * from './src/apply-labeling-error.js';
+export * from './src/from-unique-entries.js';
+export * from './src/ident-checker.js';
+export * from './src/list-difference.js';
+export * from './src/make-array-iterator.js';
+export * from './src/make-iterator.js';
+export * from './src/object-map.js';
+export * from './src/object-meta-assign.js';
+export * from './src/object-meta-map.js';
+export * from './src/throw-labeled.js';
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/common/index.js:10]: I really want to keep the content code itself under `src`. So instead I took your suggestion below. Resolving.

File: packages/common/package.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    "url": "git+https://github.com/endojs/endo.git"
+  },
+  "bugs": {
+    "url": "https://github.com/endojs/endo/issues"
+  },
+  "type": "module",
+  "main": "./index.js",
+  "module": "./index.js",
+  "exports": {
+    ".": "./index.js",
```
[the remark this answers is not in the document]
Comment (erights) [reply packages/common/package.json:21]: Done!

I gotta say though, this is a maintenance hazard. Too easy for someone to add a new one and forget to update these.

But resolving.

File: packages/common/src/ident-checker.js
[the remark this answers is not in the document]
Comment (erights) [reply packages/common/src/ident-checker.js:1]: I left it as is. Resolving.

File: packages/utils/README.md
Comment (erights) [reply packages/utils/README.md:1]: Done

**Pull request merged**